File: app/src/main/java/uk/co/sundroid/activity/data/fragments/AbstractTimeFragment.java
package uk.co.sundroid.activity.data.fragments;

import android.os.Bundle;
import android.view.*;
import android.view.View.OnClickListener;
import android.view.View.OnTouchListener;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import uk.co.sundroid.R;
import uk.co.sundroid.R.id;
import uk.co.sundroid.activity.data.fragments.dialogs.date.DatePickerFragment;
import uk.co.sundroid.activity.data.fragments.dialogs.date.TimePickerFragment;
import uk.co.sundroid.domain.LocationDetails;
import uk.co.sundroid.util.SharedPrefsHelper;
import uk.co.sundroid.util.StringUtils;
import uk.co.sundroid.util.view.ButtonDragGestureDetector;
import uk.co.sundroid.util.view.ButtonDragGestureDetector.ButtonDragGestureDetectorListener;
import uk.co.sundroid.util.LogWrapper;
import uk.co.sundroid.util.time.TimeUtils;
import uk.co.sundroid.util.time.Time;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

public abstract class AbstractTimeFragment extends AbstractDataFragment implements OnClickListener, OnTouchListener, OnSeekBarChangeListener, DatePickerFragment.OnDateSelectedListener, TimePickerFragment.OnTimeSelectedListener {
	
	private static final String TAG = AbstractTimeFragment.class.getSimpleName();

    private SimpleDateFormat dateFormat = new SimpleDateFormat("dd MMM yyyy", Locale.US);
    private SimpleDateFormat weekdayFormat = new SimpleDateFormat("EEEE", Locale.US);

    private GestureDetector dateDetector;
    private GestureDetector timeDetector;

    protected abstract int getLayout();

    protected abstract void initialise(final LocationDetails location, final Calendar dateCalendar, final Calendar timeCalendar, final View view) throws Exception;

    protected abstract void update(final LocationDetails location, final Calendar dateCalendar, final Calendar timeCalendar, final View view, boolean timeOnly) throws Exception;

    @Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle state) {
		if (container == null) {
			return null;
		}
		View view = inflater.inflate(getLayout(), container, false);
        safeInitialise(view);
		safeUpdate(view, false);
		return view;
	}

    @Override
    public void initialise() {
        if (getView() != null) {
            safeInitialise(getView());
        }
    }

    @Override
    public void update() {
        if (getView() != null) {
            safeUpdate(getView(), false);
        }
    }

    @Override
    public void onDateSet(int year, int month, int date) {
        getDateCalendar().set(year, month, date);
        getTimeCalendar().set(year, month, date);
        update();
    }

    @Override
    public void onTimeSet(int hour, int minute) {
        getTimeCalendar().set(Calendar.HOUR_OF_DAY, hour);
        getTimeCalendar().set(Calendar.MINUTE, minute);
        update();
    }

    @Override
    public void onClick(View button) {
        switch (button.getId()) {
            case id.timeButton:
                showTimePicker();
                return;
            case id.dateButton:
                showDatePicker();
                return;
            case id.zoneButton:
                startTimeZone();
                return;
        }
    }

    public boolean onTouch(View view, MotionEvent event) {
        if (view.getId() == id.dateButton) {
            return dateDetector.onTouchEvent(event);
        } else if (view.getId() == id.timeButton) {
            return timeDetector.onTouchEvent(event);
        }
        return false;
    }

    private void safeInitialise(View view) {
        LocationDetails location = getLocation();
        Calendar dateCalendar = getDateCalendar();
        Calendar timeCalendar = getTimeCalendar();
        try {
            if (location != null && dateCalendar != null && timeCalendar != null && view != null && !isDetached()) {
                initGestures(view);
                updateDateAndTime(view, dateCalendar, timeCalendar);
                initialise(location, dateCalendar, timeCalendar, view);
            }
        } catch (Exception e) {
            LogWrapper.e(TAG, "Failed to init data view", e);
        }
    }

	private void safeUpdate(View view, boolean timeOnly) {
		LocationDetails location = getLocation();
        Calendar dateCalendar = getDateCalendar();
        Calendar timeCalendar = getTimeCalendar();
		try {
			if (location != null && dateCalendar != null && timeCalendar != null && view != null && !isDetached()) {
                updateDateAndTime(view, dateCalendar, timeCalendar);
				update(location, dateCalendar, timeCalendar, view, timeOnly);
            }
		} catch (Exception e) {
			LogWrapper.e(TAG, "Failed to update data view", e);
		}
	}

    private void initGestures(View view) {
        if (dateDetector == null) {
            ButtonDragGestureDetectorListener dateListener = new ButtonDragGestureDetectorListener() {
                @Override public void onButtonDragUp() { prevMonth(); }
                @Override public void onButtonDragDown() { nextMonth(); }
                @Override public void onButtonDragLeft() { prevDate(); }
                @Override public void onButtonDragRight() { nextDate(); }
            };
            dateDetector = new GestureDetector(getApplicationContext(), new ButtonDragGestureDetector(dateListener, getApplicationContext()));
        }
        view.findViewById(id.dateButton).setOnClickListener(this);
        view.findViewById(id.dateButton).setOnTouchListener(this);
        view.findViewById(id.zoneButton).setOnClickListener(this);

        if (timeDetector == null) {
            ButtonDragGestureDetectorListener timeListener = new ButtonDragGestureDetectorListener() {
                @Override public void onButtonDragUp() { prevHour(); }
                @Override public void onButtonDragDown() { nextHour(); }
                @Override public void onButtonDragLeft() { prevMinute(); }
                @Override public void onButtonDragRight() { nextMinute(); }
            };
            timeDetector = new GestureDetector(getApplicationContext(), new ButtonDragGestureDetector(timeListener, getApplicationContext()));
        }
        view.findViewById(id.timeButton).setOnClickListener(this);
        view.findViewById(id.timeButton).setOnTouchListener(this);
        ((SeekBar)view.findViewById(R.id.timeSeeker)).setOnSeekBarChangeListener(this);
    }

    private void updateDateAndTime(View view, Calendar dateCalendar, Calendar timeCalendar) {
        LocationDetails location = getLocation();
        if (location == null || view == null) {
            return;
        }


    	if (SharedPrefsHelper.getShowTimeZone(getApplicationContext())) {
    		showInView(view, id.zoneButton);
			TimeZone zone = location.getTimeZone().getZone();
			boolean zoneDST = zone.inDaylightTime(new Date(dateCalendar.getTimeInMillis() + (12 * 60 * 60 * 1000)));
			String zoneName = zone.getDisplayName(zoneDST, TimeZone.LONG);
			textInView(view, id.zoneName, zoneName);

			String zoneCities = location.getTimeZone().getOffset(dateCalendar.getTimeInMillis() + (12 * 60 * 60 * 1000)); // Get day's main offset.
			if (StringUtils.isNotEmpty(location.getTimeZone().getCities())) {
				zoneCities += " " + location.getTimeZone().getCities();
			}
			textInView(view, id.zoneCities, zoneCities);
    	} else {
    		removeInView(view, id.zoneButton);
    	}

        Time time = TimeUtils.formatTime(getApplicationContext(), timeCalendar, false, false);
        showInView(view, id.timeHM, time.getTime() + time.getMarker());

        dateFormat.setTimeZone(dateCalendar.getTimeZone());
		weekdayFormat.setTimeZone(dateCalendar.getTimeZone());
		String date = dateFormat.format(new Date(dateCalendar.getTimeInMillis()));
		String weekday = weekdayFormat.format(new Date(dateCalendar.getTimeInMillis()));
		showInView(view, id.dateDMY, date);
		showInView(view, id.dateWeekday, weekday);

        int minutes = (timeCalendar.get(Calendar.HOUR_OF_DAY) * 60) + timeCalendar.get(Calendar.MINUTE);
        ((SeekBar)view.findViewById(R.id.timeSeeker)).setProgress(minutes);
    }

    private void showDatePicker() {
        DatePickerFragment datePickerFragment = DatePickerFragment.newInstance(getDateCalendar());
        datePickerFragment.setTargetFragment(this, 0);
        datePickerFragment.show(getFragmentManager(), "datePicker");
    }

    private void showTimePicker() {
        TimePickerFragment timePickerFragment = TimePickerFragment.newInstance(getTimeCalendar());
        timePickerFragment.setTargetFragment(this, 0);
        timePickerFragment.show(getFragmentManager(), "timePicker");
    }

    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        if (fromUser && getTimeCalendar() != null) {
            int hours = progress/60;
            int minutes = progress - (hours * 60);
            getTimeCalendar().set(Calendar.HOUR_OF_DAY, hours);
            getTimeCalendar().set(Calendar.MINUTE, minutes);
            safeUpdate(getView(), false);
        }
    }

    public void onStartTrackingTouch(SeekBar seekBar) {
        // No action.
    }

    public void onStopTrackingTouch(SeekBar seekBar) {
        // No action.
    }

    private void nextDate() {
        getDateCalendar().add(Calendar.DAY_OF_MONTH, 1);
        getTimeCalendar().add(Calendar.DAY_OF_MONTH, 1);
        safeUpdate(getView(), false);
    }

    private void nextMonth() {
        getDateCalendar().add(Calendar.MONTH, 1);
        getTimeCalendar().add(Calendar.MONTH, 1);
        safeUpdate(getView(), false);
    }

    private void prevDate() {
        getDateCalendar().add(Calendar.DAY_OF_MONTH, -1);
        getTimeCalendar().add(Calendar.DAY_OF_MONTH, -1);
        safeUpdate(getView(), false);
    }

    private void prevMonth() {
        getDateCalendar().add(Calendar.MONTH, -1);
        getTimeCalendar().add(Calendar.MONTH, -1);
        safeUpdate(getView(), false);
    }

    private void nextMinute() {
        if (getView() != null) {
            int dayOfYear = getTimeCalendar().get(Calendar.DAY_OF_YEAR);
            getTimeCalendar().add(Calendar.MINUTE, 1);
            int minutes = (getTimeCalendar().get(Calendar.HOUR_OF_DAY) * 60) + getTimeCalendar().get(Calendar.MINUTE);
            ((SeekBar)getView().findViewById(id.timeSeeker)).setProgress(minutes);
            if (getTimeCalendar().get(Calendar.DAY_OF_YEAR) != dayOfYear) {
                getDateCalendar().add(Calendar.DAY_OF_MONTH, 1);
                safeUpdate(getView(), false);
            } else {
                safeUpdate(getView(), true);
            }
        }
    }

    private void prevMinute() {
        if (getView() != null) {
            int dayOfYear = getTimeCalendar().get(Calendar.DAY_OF_YEAR);
            getTimeCalendar().add(Calendar.MINUTE, -1);
            int minutes = (getTimeCalendar().get(Calendar.HOUR_OF_DAY) * 60) + getTimeCalendar().get(Calendar.MINUTE);
            ((SeekBar)getView().findViewById(id.timeSeeker)).setProgress(minutes);
            if (getTimeCalendar().get(Calendar.DAY_OF_YEAR) != dayOfYear) {
                getDateCalendar().add(Calendar.DAY_OF_MONTH, -1);
                safeUpdate(getView(), false);
            } else {
                safeUpdate(getView(), true);
            }
        }
    }

    private void nextHour() {
        if (getView() != null) {
            int dayOfYear = getTimeCalendar().get(Calendar.DAY_OF_YEAR);
            getTimeCalendar().add(Calendar.HOUR_OF_DAY, 1);
            int minutes = (getTimeCalendar().get(Calendar.HOUR_OF_DAY) * 60) + getTimeCalendar().get(Calendar.MINUTE);
            ((SeekBar)getView().findViewById(id.timeSeeker)).setProgress(minutes);
            if (getTimeCalendar().get(Calendar.DAY_OF_YEAR) != dayOfYear) {
                getDateCalendar().add(Calendar.DAY_OF_MONTH, 1);
                safeUpdate(getView(), false);
            } else {
                safeUpdate(getView(), true);
            }
        }
    }

    private void prevHour() {
        if (getView() != null) {
            int dayOfYear = getTimeCalendar().get(Calendar.DAY_OF_YEAR);
            getTimeCalendar().add(Calendar.HOUR_OF_DAY, -1);
            int minutes = (getTimeCalendar().get(Calendar.HOUR_OF_DAY) * 60) + getTimeCalendar().get(Calendar.MINUTE);
            ((SeekBar)getView().findViewById(id.timeSeeker)).setProgress(minutes);
            if (getTimeCalendar().get(Calendar.DAY_OF_YEAR) != dayOfYear) {
                getDateCalendar().add(Calendar.DAY_OF_MONTH, -1);
                safeUpdate(getView(), false);
            } else {
                safeUpdate(getView(), true);
            }
        }
    }
    
}

File: app/src/main/kotlin/uk/co/sundroid/activity/data/fragments/AbstractTimeFragment.kt
package uk.co.sundroid.activity.data.fragments

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.GestureDetector
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.DatePicker
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TimePicker
import kotlinx.android.synthetic.main.inc_timebar.*
import kotlinx.android.synthetic.main.inc_timebar.dateButton
import kotlinx.android.synthetic.main.inc_timebar.dateDMY
import kotlinx.android.synthetic.main.inc_timebar.dateWeekday
import kotlinx.android.synthetic.main.inc_timebar.zoneButton
import uk.co.sundroid.activity.Page
import uk.co.sundroid.util.time.formatTimeStr
import uk.co.sundroid.util.view.ButtonDragGestureDetector
import uk.co.sundroid.util.view.ButtonDragGestureDetector.ButtonDragGestureDetectorListener
import java.text.SimpleDateFormat
import java.util.*
import java.util.Calendar.DAY_OF_MONTH
import java.util.Calendar.MONTH
import java.util.Calendar.MINUTE
import java.util.Calendar.HOUR_OF_DAY
import java.util.Calendar.YEAR

abstract class AbstractTimeFragment : AbstractDataFragment(), OnSeekBarChangeListener {

    private val dateFormat = SimpleDateFormat("dd MMM yyyy", Locale.US)
    private val weekdayFormat = SimpleDateFormat("EEEE", Locale.US)

    protected abstract val layout: Int

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, state: Bundle?): View? {
        return inflater.inflate(layout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        initGestures()
        updateDate()
        updateTime()
        initialise()
        updateData(view, false)
    }

    override fun update(dateChanged: Boolean, timeChanged: Boolean) {
        val view = view
        if (isSafe && view != null) {
            initGestures()
            updateDate()
            updateTime()
            updateData(view, !dateChanged)
        }
    }

    private fun onDateSet(year: Int, month: Int, date: Int) {
        getDateCalendar().set(year, month, date)
        getTimeCalendar().set(year, month, date)
        update(dateChanged = true, timeChanged = false)
    }

    private fun onTimeSet(hour: Int, minute: Int) {
        getTimeCalendar().set(HOUR_OF_DAY, hour)
        getTimeCalendar().set(MINUTE, minute)
        update(dateChanged = false, timeChanged = true)
    }

    private fun initGestures() {
        val dateListener = object : ButtonDragGestureDetectorListener {
            override fun onButtonDragUp() = calendarDiff(MONTH, -1)
            override fun onButtonDragDown() = calendarDiff(MONTH, 1)
            override fun onButtonDragLeft() = calendarDiff(DAY_OF_MONTH, -1)
            override fun onButtonDragRight() = calendarDiff(DAY_OF_MONTH, 1)
        }
        val dateDetector = GestureDetector(requireContext(), ButtonDragGestureDetector(dateListener, requireContext()))
        zoneButton.setOnClickListener { setPage(Page.TIME_ZONE) }
        dateButton.setOnClickListener { showDatePicker() }
        dateButton.setOnTouchListener { _, e -> dateDetector.onTouchEvent(e) }

        val timeListener = object : ButtonDragGestureDetectorListener {
            override fun onButtonDragUp() = calendarDiff(HOUR_OF_DAY, -1)
            override fun onButtonDragDown() = calendarDiff(HOUR_OF_DAY, 1)
            override fun onButtonDragLeft() = calendarDiff(MINUTE, -1)
            override fun onButtonDragRight() = calendarDiff(MINUTE, 1)
        }
        val timeDetector = GestureDetector(requireContext(), ButtonDragGestureDetector(timeListener, requireContext()))
        timeButton.setOnClickListener { showTimePicker() }
        timeButton.setOnTouchListener { _, e -> timeDetector.onTouchEvent(e) }
        timeSeeker.setOnSeekBarChangeListener(this)
    }

    private fun updateDate() {
        updateTimeZone()
        val calendar = getDateCalendar()
        dateFormat.timeZone = calendar.timeZone
        weekdayFormat.timeZone = calendar.timeZone
        val date = dateFormat.format(Date(calendar.timeInMillis))
        val weekday = weekdayFormat.format(Date(calendar.timeInMillis))
        dateDMY.text = date
        dateWeekday.text = weekday
    }

    private fun updateTime() {
        val calendar = getTimeCalendar()
        val time = formatTimeStr(requireContext(), calendar, allowSeconds = false, allowRounding = false)
        timeHM.text = time
        timeSeeker.progress = calendar.get(HOUR_OF_DAY) * 60 + calendar.get(MINUTE)
    }

    private fun showDatePicker() {
        val calendar = getDateCalendar()
        val today = Calendar.getInstance(calendar.timeZone)
        val listener = { _: DatePicker, y: Int, m: Int, d: Int -> onDateSet(y, m, d) }
        val dialog = DatePickerDialog(requireContext(), listener, calendar.get(YEAR), calendar.get(MONTH), calendar.get(DAY_OF_MONTH))
        dialog.setButton(DialogInterface.BUTTON_NEUTRAL, "Today") { _, _ -> onDateSet(today.get(YEAR), today.get(MONTH), today.get(DAY_OF_MONTH))}
        dialog.show()
    }

    private fun showTimePicker() {
        val calendar = getTimeCalendar()
        val now = Calendar.getInstance(calendar.timeZone)
        val listener = { _: TimePicker, h: Int, m: Int -> onTimeSet(h, m) }
        val dialog = TimePickerDialog(activity, listener, calendar.get(HOUR_OF_DAY), calendar.get(MINUTE), prefs.clockType24())
        dialog.setButton(DialogInterface.BUTTON_NEUTRAL, "Now") { _, _ -> onTimeSet(now.get(HOUR_OF_DAY), now.get(MINUTE))}
        dialog.show()
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        if (fromUser) {
            val hour = progress / 60
            val minute = progress - hour * 60
            onTimeSet(hour, minute)
        }
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) { }
    override fun onStopTrackingTouch(seekBar: SeekBar) { }

    protected abstract fun updateData(view: View, timeOnly: Boolean)

}
